D91-D93 CLOUDS2 helper instrumentation (PRECIP_MP, ANVIL_OPTICAL_THICKNESS, MC_CLOUD_FRACTION, CONVECTIVE_MICROPHYSICS, MC_PRECIP_PHASE, MASS_FLUX). Generated by diff -u against the PRISTINE source (2026-10-05); independent of the D89 CLOUDS2 patch (verified to apply to pristine and after CLOUDS2_dq.f90.patch).

--- CLOUDS2.F90.orig
+++ CLOUDS2.F90
@@ -5324,6 +5324,7 @@
     TAUMC=1.5*TEM/(FCLD*RCLDE+1.E-20)
     if(TAUMC.gt.100.) TAUMC=100.
 
+    call ffh_rec(2,1,12,(/SVLATL,RCLDLX,RCLDIX,MCDNCW,MCDNCI,RIMAX,BYBR,FCLD,TEM,WTEM,RCLD,TAUMC/))
     return
   end subroutine ANVIL_OPTICAL_THICKNESS
 
@@ -5388,6 +5389,8 @@
     end if
     if(FCLOUD.gt.1.) FCLOUD=1.
 
+    call ffh_rec(3,1,15,(/dble(L),TL,PL,CCM,WCU,PLEMIN,PLEL2,PLEMAX,CCMMIN,WCUMIN,dble(LMIN),dble(LMAX), &
+         CCMUL,CCMUL2,FCLOUD/))
     return
   end subroutine MC_CLOUD_FRACTION
 
@@ -5443,6 +5446,7 @@
 !@var CONDIP,CONDGP precipitating part of convective condensate
     integer ITER
 !@var ITER index for iteration
+    real*8 :: ffcip0,ffcgp0
 #ifdef CLD_AER_CDNC
     real*8 TL,RCLD_C,MCDNCW,CONDMU,CONDPC,RHO
 !@var TL temperature
@@ -5465,6 +5469,7 @@
     !**** ASSUMED TO DETRAIN AT THE CURRENT LEVEL; FIRST FIND SMALL PARTICLE END OF
     !**** THIS RANGE (LOWER CRITICAL SIZE) FOR GRAUPEL AND FLUFFY ICE
 
+    ffcip0=CONDIP ; ffcgp0=CONDGP
     WV=WCU-DWCU          ! apply the calculated cumulus updraft speed
     if(WV.lt.0.d0) WV=0.d0
 
@@ -5501,6 +5506,7 @@
       end do
 
       !**** PART OF LIQUID CONDENSATE NOT ADVECTED UP (EQ. 5, DEL GENIO ET AL. (2005))
+      call ffh_site(1)
       CONDP1 = PRECIP_MP (RHOW,FLAMW,DCW,CN0)
 
       !**** UPPER CRITICAL SIZE FOR LIQUID WATER
@@ -5515,6 +5521,7 @@
       end do
 
       !**** PRECIPITATING PART OF LIQUID CONDENSATE (RAIN)
+      call ffh_site(2)
       CONDP = PRECIP_MP (RHOW,FLAMW,DCW,CN0)
 
 #ifndef NO_CLD_AER_CDNC_IN_MCMICRO
@@ -5538,6 +5545,7 @@
 
       !**** ABOVE MIXED PHASE REGION, PART OF FLUFFY ICE NOT ADVECTED UP
     else if (TP.le.TIG) then ! pure ice phase, use TIG
+      call ffh_site(3)
       CONDP1 = PRECIP_MP (RHOIP,FLAMI,DCI,CN0I)
 
       !**** NOW FIND LARGE PARTICLE END OF RANGE FOR DETRAINMENT AT CURRENT
@@ -5549,6 +5557,7 @@
       DCI=min(DCI,1.D-2)
 
       !**** PRECIPITATING PART OF FLUFFY ICE (SNOW)
+      call ffh_site(4)
       CONDP = PRECIP_MP (RHOIP,FLAMI,DCI,CN0I)
 
       !**** WITHIN MIXED PHASE REGION, CALCULATE FRACTION OF FROZEN CONDENSATE
@@ -5561,7 +5570,9 @@
       FI=1.-FG
 
       !**** PART OF FROZEN CONDENSATE NOT ADVECTED UP IN MIXED PHASE REGION
+      call ffh_site(5)
       CONDIP = PRECIP_MP (RHOIP,FLAMI,DCI,CN0I)
+      call ffh_site(6)
       CONDGP = PRECIP_MP (RHOG, FLAMG,DCG,CN0G)
       CONDP1=FG*CONDGP+FI*CONDIP
       WV=WCU+DWCU
@@ -5575,11 +5586,16 @@
       DCG=min(DCG,1.D-2)
 
       !**** PRECIPITATING PART OF FROZEN CONDENSATE (HAIL + SNOW) IN MIXED PHASE REGION
+      call ffh_site(7)
       CONDIP = PRECIP_MP (RHOIP,FLAMI,DCI,CN0I)
+      call ffh_site(8)
       CONDGP = PRECIP_MP (RHOG, FLAMG,DCG,CN0G)
       CONDP=FG*CONDGP+FI*CONDIP
     end if
 
+    call ffh_rec(4,1,27,(/PL,WCU,DWCU,dble(LFRZ),WCUFRZ,TP,TI,FITMAX,PLAND,CN0,CN0I,CN0G, &
+         FLAMW,FLAMG,FLAMI,RHOIP,RHOG,dble(ITMAX),TLMIN,TLMIN1,WMAX,ffcip0,ffcgp0, &
+         CONDP,CONDP1,CONDIP,CONDGP/))
     return
   end subroutine CONVECTIVE_MICROPHYSICS
 
@@ -5614,8 +5630,10 @@
     real*8, intent(OUT) :: HEAT1
 !@var PRCP precipitation
     real*8, intent(INOUT) :: PRCP
+    real*8 :: ffh1prcp,ffh1heat
 
     !**** DETERMINE THE AIR MASS FOR RAIN EVAPORATION TO TAKE PLACE
+    ffh1prcp=PRCP ; ffh1heat=HEAT1
     MCLOUD=0.
 !**** Option to revert to AR5 approach of only allowing re-evaporation
 !**** below cloud base.
@@ -5650,6 +5668,8 @@
            )*COND*BYSHA
     end if
 
+    call ffh_rec(5,1,16,(/dble(L),AIRM,FEVAP,LHP1,ffh1prcp,TOLD,TOLD1,VLAT,COND,dble(LMIN), &
+         ffh1heat,dble(MC_REVP_ABV_CLDBASE),PRCP,LHP,MCLOUD,HEAT1/))
     return
   end subroutine MC_PRECIP_PHASE
 
@@ -5670,11 +5690,14 @@
     implicit none
     !
     real*8   PRECIP_MP, RHO, FLAM, DC, CN
+    integer :: ffh_cur
+    common /ffhsite/ ffh_cur
     !
     PRECIP_MP = RHO*(PI*BY6)*CN*exp(-FLAM*DC)* &
          (DC*DC*DC/FLAM+3.*DC*DC/(FLAM*FLAM)+ &
          6.*DC/(FLAM*FLAM*FLAM)+6./FLAM**4.)
     !
+    call ffh_rec(1,ffh_cur,5,(/RHO,FLAM,DC,CN,PRECIP_MP/))
     return
   end function PRECIP_MP
 
@@ -5799,6 +5822,10 @@
       if(DMSE1.lt.-1.d-3) FPLUME=FPLUME+DFP
     end do
     !
+    call ffh_rec(6,1,28,(/dble(LMIN),LHX,QMO1,QMO2,SMO1,SMO2,SLH,WMDN,WMUP,WMEDG, &
+         AIRM(LMIN),AIRM(LMIN+1),BYAM(LMIN),BYAM(LMIN+1),BYAM(LMIN+2),SM(LMIN+2),QM(LMIN+2), &
+         PLK(LMIN),PLK(LMIN+1),PL(LMIN),PL(LMIN+1), &
+         FPLUME,FMP2,DQSUM,dble(ITER),DMSE1,TNX,QNX/))
     return
     !
   end subroutine MASS_FLUX
